northstar/src/runtime/repository.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2021 ESRLabs
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+    time::SystemTime,
+};
+
+use super::{error::Error, ContainerKey, Name, RepositoryId, Version};
+use ed25519_dalek::PublicKey;
+use log::debug;
+use npk::npk::Npk;
+use tokio::fs;
+
+#[derive(Debug, Clone)]
+pub(super) struct Repository {
+    pub(super) id: RepositoryId,
+    pub(super) dir: PathBuf,
+    pub(super) key: Option<PublicKey>,
+    last_modified: Option<SystemTime>,
+    containers: HashMap<(Name, Version), PathBuf>,
+}
+
+impl Repository {
+    pub async fn new(
+        id: RepositoryId,
+        dir: PathBuf,
+        key: Option<PublicKey>,
+    ) -> Result<Repository, Error> {
+        let mut repository = Repository {
+            id,
+            dir,
+            key,
+            last_modified: None,
+            containers: HashMap::new(),
+        };
+        repository.scan().await?;
+        Ok(repository)
+    }
+
+    pub fn npks(&self) -> &HashMap<(Name, Version), PathBuf> {
+        &self.containers
+    }
+
+    pub async fn add(&mut self, key: &ContainerKey, src: &Path) -> Result<(), Error> {
+        let dest = self
+            .dir
+            .join(format!("{}-{}.npk", key.name(), key.version()));
```

Comment:
adding a container to a repository should also update the `last_modified` timestamp

Reply:
Yes. Added.